npe2/_plugin_manager.py

```diff
@@ -52,105 +54,220 @@ def addi(self, begin: int, end: int, data: T) -> None:
+
 
 __all__ = ["PluginContext", "PluginManager"]
 PluginName = str  # this is `PluginManifest.name`
 
 
 class _ContributionsIndex:
     def __init__(self) -> None:
-        self._submenus: Dict[str, SubmenuContribution] = {}
+        self._indexed: Set[str] = set()
         self._commands: Dict[str, Tuple[CommandContribution, PluginName]] = {}
-        self._themes: Dict[str, ThemeContribution] = {}
-        self._widgets: List[WidgetContribution] = []
         self._readers: DefaultDict[str, List[ReaderContribution]] = DefaultDict(list)
-        self._samples: DefaultDict[str, List[SampleDataContribution]] = DefaultDict(
-            list
-        )
-        self._writers_by_type: DefaultDict[
+        self._writers: DefaultDict[
             LayerType, TypedIntervalTree[WriterContribution]
         ] = DefaultDict(IntervalTree)
-        self._writers_by_command: DefaultDict[
-            str, List[WriterContribution]
-        ] = DefaultDict(list)
-
-    def index_contributions(self, ctrb: ContributionPoints, key: PluginName):
-        if ctrb.sample_data:
-            self._samples[key] = ctrb.sample_data
-        for cmd in ctrb.commands or []:
-            self._commands[cmd.id] = cmd, key
-        for subm in ctrb.submenus or []:
-            self._submenus[subm.id] = subm
-        for theme in ctrb.themes or []:
-            self._themes[theme.id] = theme
-        self._widgets.extend(ctrb.widgets or [])
-        for reader in ctrb.readers or []:
+
+    def index_contributions(self, manifest: PluginManifest):
+        ctrb = manifest.contributions
+        if not ctrb or manifest.name in self._indexed:
+            return  # pragma: no cover
+
+        self._indexed.add(manifest.name)
+        for cmd in ctrb.commands or ():
+            self._commands[cmd.id] = cmd, manifest.name
+        for reader in ctrb.readers or ():
             for pattern in reader.filename_patterns:
                 self._readers[pattern].append(reader)
             if reader.accepts_directories:
                 self._readers[""].append(reader)
-        for writer in ctrb.writers or []:
-            self._writers_by_command[writer.command].append(writer)
+        for writer in ctrb.writers or ():
             for c in writer.layer_type_constraints():
-                self._writers_by_type[c.layer_type].addi(*c.bounds, writer)
+                self._writers[c.layer_type].addi(*c.bounds, writer)
+
+    def remove_contributions(self, key: PluginName) -> None:
+        """This must completely remove everything added by `index_contributions`."""
+        if key not in self._indexed:
+            return  # pragma: no cover
+
+        for cmd_id, (_, plugin) in list(self._commands.items()):
+            if key == plugin:
+                del self._commands[cmd_id]
+        for contribs in self._readers.values():
+            for i, ctrb in reversed(list(enumerate(contribs))):
+                if ctrb.plugin_name == key:
+                    contribs.pop(i)
```

Comment:
Aren't we starting to make things really complicated using dicts, instead of using something flatter, and filtering in the relevant methods ?

Reply:
perhaps!  not married at all to this one if you want to propose a better structure

Reply:
Ok, let' try to finish that and I'll suggest a refactor.